nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/source/DirConfigSource.java

```diff
@@ -0,0 +1,377 @@
+        // only delete from the map if the file that has been deleted is the same as the one stored in the map
+        // -> deleting a file less specific but matching a property should not remove from the map
+        // -> deleting a file more specific than in map shouldn't occur (it had to slip through longest match check then).
+        if (path.equals(properties.get(property).path)) {
+            properties.remove(property);
+        }
+        
+    }
+    
+    String parsePropertyNameFromPath(Path path) {
+        // 1. get relative path based on the config dir ("/config"),
+        String property = directory.relativize(path.getParent()).toString();
+        // 2. ignore all file suffixes after last dot
+        property += path.getFileName().toString().substring(0, path.getFileName().toString().lastIndexOf('.')-1);
+        // 3. replace all path seps with a ".",
+        property = property.replace(File.separatorChar, '.');
+        // so "/config/foo/bar/test/one.txt" becomes "foo/bar/test/one.txt" becomes "foo.bar.test.one" property name
+        return property;
+    }
+    
+    /**
+     * Check if the path given is a more specific path to a value for the given property
+     * @param property
+     * @param path
+     * @return true if more specific, false if not
+     */
+    boolean checkLongestMatchForPath(String property, Path path) {
+        // Make path relative to config directory
+        // NOTE: we will never have a path containing "..", as our tree walkers are always inside this "root".
+        Path relativePath = directory.relativize(path);
+        
+        // No property -> path is new and more specific
+        if (! properties.containsKey(property))
+            return true;
+        DirProperty old = properties.get(property);
+        
+        // Check if this element has a higher path depth (longest match)
+        // Example: "foo.bar/test/one.txt" (depth 2) wins over "foo.bar.test.one.txt" (depth 0)
+        boolean depth = old.pathDepth > relativePath.getNameCount();
+        
+        // In case that both pathes have the same depth, we need to check on the position of dots.
+        // Example: /config/foo.bar/test/one.txt is less specific than /config/foo/bar.test/one.txt
+        if (old.pathDepth == relativePath.getNameCount()) {
+            String oldPath = old.path.toString();
+            String newPath = path.toAbsolutePath().toString();
+            int offset = 0;
+            while (offset > -1) {
+                if (newPath.indexOf(".", offset) > oldPath.indexOf(".", offset)) return true;
+                offset = oldPath.indexOf(".", offset + 1);
+            }
+        }
+        return depth;
+    }
+    
+    DirProperty readPropertyFromPath(Path path, BasicFileAttributes mainAtts) throws IOException {
+        if (Files.exists(path) && Files.isRegularFile(path) && Files.isReadable(path)) {
+            return new DirProperty(
+                new String(Files.readAllBytes(path)),
```

Comment:
I think you want to specify the encoding explicitly here.

Reply:
Using UTF-8.